### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by mattruggio**
This PR includes the initial implementation, tests, and documentation.  Note that this library is the 'core' modeling library and really is not meant to be used by itself.  I will link the other corresponding library to this PR once it is ready.

### CODE REVIEW

**Comment by mattruggio** [comment]
One interesting thing to note here is my initial implementation of Column definitions.  I only included 2: Undefined and Boolean.  I am on the fence if this should even be a concern of this library, but my gut tells me yes.  Basically what a Column definition does is provide value coercion to ensure the inputted value conforms the desired type.  For our immediate use we only need Boolean, but I am assuming we could/should/would eventually support more (such as Date, DateTime, Integer, etc.) . Another idea would be how this could be inferred through reflection, but the problem is you are now coupling in DBMS notions, which this library stays away from.  Maybe it becomes a concern strictly for the SQL generator plugin?

In any event, I kept it to at least provide a pattern for adding more types but I am fully aware more may need to be added if/when necessary.

File: lib/dbee/query.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def initialize(fields: [], filters: [], limit: nil, sorters: [])
+      @fields   = Field.array(fields)
+      @filters  = Filters.array(filters)
+      @limit    = limit.to_s.empty? ? nil : limit.to_i
+      @sorters  = Sorter.array(sorters)
+
+      freeze
+    end
+
+    def ==(other)
```
Comment (gerryster) [inline lib/dbee/query.rb]: One of these methods can be an alias.

File: lib/dbee/query/key_path.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    class KeyPath
+      extend Forwardable
+
+      class << self
+        def get(obj)
+          obj.is_a?(self.class) ? obj : new(obj)
+        end
+      end
+
+      SPLIT_CHAR = '.'
```
Comment (gerryster) [inline lib/dbee/query/key_path.rb]: Maybe make this private?

File: lib/dbee/base.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+      def to_model(name = nil, constraints = [])
+        name  = derive_name(name)
+        key   = [name, constraints]
+
+        to_models[key] ||= Model.make(model_config(name, constraints))
+      end
+
+      def table_name
+        @table_name.to_s
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/dbee/base.rb]: This takes care of the case where one has not been set (== nil)

[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/dbee/base.rb]: I am going to settle on: `@table_name || ''` . I think this way describes the base case more clearly.

File: lib/dbee/base.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      def columns_by_name
+        @columns_by_name ||= {}
+      end
+
+      def associations_by_name
+        @associations_by_name ||= {}
+      end
+
+      def inherited_table_name
+        subclasses.each do |subclass|
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/dbee/base.rb]: I changed it to this, remember I don't have ActiveModel's present? monkeypatch.

````ruby
def table_name?
  !table_name.empty?
end

def inherited_table_name
  subclasses.find(&:table_name?)&.table_name || ''
end
````

File: lib/dbee/model.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+require_relative 'model/columns'
+require_relative 'model/constraints'
+
+module Dbee
+  # In DB terms, a Model is usually a table, but it does not have to be.  You can also re-model
+  # your DB schema using Dbee::Models.
+  class Model
+    acts_as_hashable
+
+    JOIN_CHAR = '.'
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/dbee/model.rb]: Yes it can be.

File: lib/dbee/model/columns/boolean.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+        private
+
+        def null_or_empty?(val)
+          val.nil? || val.to_s.empty?
+        end
+
+        # rubocop:disable Style/DoubleNegation
+        def nully?(val)
+          null_or_empty?(val) || !!(val.to_s =~ /(nil|null)$/i)
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/dbee/model/columns/boolean.rb]: The reason I did not use String#match? was it was not supported in Ruby 2.3.8.  I updated this library and will update the corresponding other ActiveRecord library to be minimum 2.5.3.  [I see no reason to support < 2.5.X](https://www.ruby-lang.org/en/news/2019/03/31/support-of-ruby-2-3-has-ended/).

**Comment by mattruggio** [comment]
@gerryster - Thanks for reviewing, back to you!

File: .rubocop.yml [hunk 7]
```diff
@@ -14,7 +14,7 @@ Metrics/MethodLength:
   Max: 25
 
 AllCops:
-  TargetRubyVersion: 2.3
+  TargetRubyVersion: 2.5
```
Comment (gerryster) [inline .rubocop.yml]: Do you want to target 2.6 instead?

File: lib/dbee/base.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (gerryster) [reply lib/dbee/base.rb]: Right, I forget that empty? is declared on standard library string.

File: .rubocop.yml (hunk 7 above)
Comment (gerryster) [reply .rubocop.yml]: Rubocop requires this based off of the gemspec.

**Pull request merged**